lib/shrine/plugins/derivation_endpoint.rb

```diff
@@ -0,0 +1,1295 @@
+    # Now the source `UploadedFile` will be passed as the second argument of
+    # the derivation block:
+    #
+    #     derivation :thumbnail do |file, uploaded_file, width, height|
+    #       uploaded_file             #=> #<Shrine::UploadedFile>
+    #       uploaded_file.id          #=> "9a7d1bfdad24a76f9cfaff137fe1b5c7.jpg"
+    #       uploaded_file.storage_key #=> "store"
+    #       uploaded_file.metadata    #=> {}
+    #
+    #       # ...
+    #     end
+    #
+    # By default original metadata that was extracted during attachment isn't
+    # available, to keep the derivation URL as short as possible. However, if
+    # you want to have original metadata available when deriving, you can set
+    # the `:metadata` option to the list of needed metadata values:
+    #
+    #     plugin :derivation_endpoint, metadata: ["filename", "mime_type"]
+    #
+    # Now `filename` and `mime_type` metadata values will be available in the
+    # derivation block:
+    #
+    #     derivation :thumbnail do |file, uploaded_file, width, height|
+    #       uploaded_file.metadata #=>
+    #       # {
+    #       #  "filename" => "nature.jpg",
+    #       #  "mime_type" => "image/jpeg"
+    #       # }
+    #
+    #       uploaded_file.original_filename #=> "nature.jpg"
+    #       uploaded_file.mime_type         #=> "image/jpeg"
+    #
+    #       # ...
+    #     end
+    #
+    # ## Downloading
+    #
+    # When a derivation is requested, the original uploaded file will be
+    # downloaded to disk before the derivation block is called. If you want
+    # to pass in additional storage-specific download options, you can do so
+    # via `:download_options`:
+    #
+    #     plugin :derivation_endpoint, download_options: {
+    #       sse_customer_algorithm: "AES256",
+    #       sse_customer_key:       "secret_key",
+    #       sse_customer_key_md5:   "secret_key_md5",
+    #     }
+    #
+    # When using `Shrine.derivation_endpoint` or `Shrine.derivation_response`,
+    # if the original uploaded file has been deleted, the error the storage
+    # raises when attempting to download it will be propagated by default. You
+    # can choose to have the endpoint convert these errors to 404 responses by
+    # adding them to `:download_errors`:
+    #
+    #     plugin :derivation_endpoint, download_errors: [
+    #       Errno::ENOENT,              # raised by Shrine::Storage::FileSystem
+    #       Aws::S3::Errors::NoSuchKey, # raised by Shrine::Storage::S3
+    #     ]
+    #
+    # If you don't want the uploaded file to be downloaded to disk for you, set
+    # `:download` to `false`.
+    #
+    #     plugin :derivation_endpoint, download: false
+    #
+    # In this case the `UploadedFile` object is yielded to the derivation block
+    # instead of the raw file:
+    #
+    #     derivation :thumbnail do |uploaded_file, width, height|
+    #       uploaded_file #=> #<Shrine::UploadedFile>
+    #
+    #       # ...
+    #     end
```

Comment:
What would be the use case of this?

Reply:
For example, you could call another service to do the processing, e.g [ImageOptim](https://imageoptim.com):

```rb
derivation :thumbnail do |uploaded_file, width, height|
  Down::Http.download("https://im2.io/<USERNAME>/#{width}x#{height}/#{uploaded_file.url}")
end
```

That way you could use `derivation_endpoint` plugin to wrap another on-the-fly processing service if you want (e.g. if you don't want your users to know which service you're using, or you need to tweak the behaviour).

I will add this example, thanks for pointing it out.